staging/src/k8s.io/api/events/v1/types.go

```diff
@@ -0,0 +1,119 @@
+type Event struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// eventTime is the time when this Event was first observed. It is required.
+	EventTime metav1.MicroTime `json:"eventTime" protobuf:"bytes,2,opt,name=eventTime"`
+
+	// series is data about the Event series this event represents or nil if it's a singleton Event.
+	// +optional
+	Series *EventSeries `json:"series,omitempty" protobuf:"bytes,3,opt,name=series"`
+
+	// reportingController is the name of the controller that emitted this Event, e.g. `kubernetes.io/kubelet`.
+	// This field cannot be empty for new Events and it should follow the format requirements mentioned in
+	// IsQualifiedName function under "k8s.io/apimachinery/pkg/util/validation".
+	// +optional
+	ReportingController string `json:"reportingController,omitempty" protobuf:"bytes,4,opt,name=reportingController"`
+
+	// reportingInstance is the ID of the controller instance, e.g. `kubelet-xyzf`.
+	// This field cannot be empty for new Events and it can have at most 128 characters (as suggested in ValidateEvent).
+	// +optional
+	ReportingInstance string `json:"reportingInstance,omitempty" protobuf:"bytes,5,opt,name=reportingInstance"`
+
+	// action is what action was taken/failed regarding to the regarding object. It is human-readable.
+	// This field can have at most 128 characters (as suggested in ValidateEvent).
+	// +optional
+	Action string `json:"action,omitempty" protobuf:"bytes,6,name=action"`
+
+	// reason is why the action was taken. It is human-readable.
+	// This field can have at most 128 characters (as suggested in ValidateEvent).
+	Reason string `json:"reason,omitempty" protobuf:"bytes,7,name=reason"`
+
+	// regarding contains the object this Event is about. In most cases it's an Object reporting controller
+	// implements, e.g. ReplicaSetController implements ReplicaSets and this event is emitted because
+	// it acts on some changes in a ReplicaSet object.
+	// +optional
+	Regarding corev1.ObjectReference `json:"regarding,omitempty" protobuf:"bytes,8,opt,name=regarding"`
+
+	// related is the optional secondary object for more complex actions. E.g. when regarding object triggers
+	// a creation or deletion of related object.
+	// +optional
+	Related *corev1.ObjectReference `json:"related,omitempty" protobuf:"bytes,9,opt,name=related"`
+
+	// note is a human-readable description of the status of this operation.
+	// Maximal length of the note is 1kB, but libraries should be prepared to
+	// handle values up to 64kB.
+	// +optional
+	Note string `json:"note,omitempty" protobuf:"bytes,10,opt,name=note"`
+
+	// type is the type of this event (Normal, Warning), new types could be added in the future.
+	// It is human-readable.
+	// +optional
+	Type string `json:"type,omitempty" protobuf:"bytes,11,opt,name=type"`
```

Comment:
is the value in this field validated and limited to "", "Normal", and "Warning"? I don't see it referenced in ValidateEvent

Reply:
Yes - we do validate at the reporting library level:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/events/event_recorder.go#L56

We should also add that validation at the API level as Jordan suggested.